bgpv1: Adds CiliumPodIPPool Support to PathAttr Policies

danehans opened the pull request:
- Adds the CiliumPodIPPool selector type to BGP CP CiliumBGPPathAttributes API.
- Updates RoutePolicyReconciler to support CiliumPodIPPool.
- Added unit tests for the CiliumPodIPPool selector type.
- Regenerated CiliumBGPPeeringPolicy CRD.

<!-- Description of change -->

Fixes: #28296

```release-note
Adds the CiliumPodIPPool selector type to BGP CP AdvertisedPathAttributes to match CiliumPodIPPool custom resources. Path attributes apply to routes announced for selected CiliumPodIPPools.
```

rastislavs commented on pkg/bgpv1/manager/route_policy_reconciler.go:
```diff
@@ -168,8 +176,25 @@ func (r *RoutePolicyReconciler) pathAttributesToPolicy(attrs v2alpha1api.CiliumB
+			if cidrs, ok := localPools[pool.Name]; ok {
+				for _, cidr := range cidrs {
+					if cidr.Addr().Is4() {
+						v4Prefixes = append(v4Prefixes, &types.RoutePolicyPrefixMatch{CIDR: cidr, PrefixLenMin: maxPrefixLenIPv4, PrefixLenMax: maxPrefixLenIPv4})
```
`PrefixLenMin` and `PrefixLenMax` should be equal to the actual Pool prefix length (and same for IPv6). This is most likely the reason why the policy does not apply on the advertisement.

danehans replied:
@rastislavs thanks for the guidance. ^ was indeed the issue. Commit b9cf743 resolves this conversation.

danehans committed:
bgpv1: Adds CiliumPodIPPool Support to PathAttr Policies

- Adds the CiliumPodIPPool selector type to BGP CP CiliumBGPPathAttributes API.
- Updates RoutePolicyReconciler to support CiliumPodIPPool.
- Added unit tests for the CiliumPodIPPool selector type.
- Regenerated CiliumBGPPeeringPolicy CRD.

Fixes: #28296

Signed-off-by: Daneyon Hansen <daneyon.hansen@solo.io>

rastislavs commented on pkg/bgpv1/manager/route_policy_reconciler.go:
```diff
@@ -31,16 +32,21 @@ type RoutePolicyReconcilerOut struct {
 type RoutePolicyReconciler struct {
-	ipPoolStore BGPCPResourceStore[*v2alpha1api.CiliumLoadBalancerIPPool]
+	lbStore  BGPCPResourceStore[*v2alpha1api.CiliumLoadBalancerIPPool]
+	podStore BGPCPResourceStore[*v2alpha1api.CiliumPodIPPool]
```
What about naming these `lbPoolStore` and `podPoolStore`?

danehans replied:
Commit 2085898 resolves this conversation.

rastislavs reviewed (approved):
LGTM now, thanks!

harsimran-pabla commented on pkg/bgpv1/manager/route_policy_reconciler.go:
```diff
@@ -186,8 +192,27 @@ func (r *RoutePolicyReconciler) pathAttributesToPolicy(attrs v2alpha1api.CiliumB
+			if attrs.Selector != nil && !labelSelector.Matches(labels.Set(pool.Labels)) {
+				continue
+			}
+			// only advertise pool cidrs that have been allocated to the local node.
```
Comment says 'advertise' but technically this is not to do any advertisement. It only creates egress policy.

danehans replied:
Thanks for catching this, I'll update.

harsimran-pabla commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -34,6 +34,8 @@ const (
 	// CiliumLoadBalancerIPPoolSelectorName defines the name for a selector matching CiliumLoadBalancerIPPool resources.
 	CiliumLoadBalancerIPPoolSelectorName = "CiliumLoadBalancerIPPool"
+	// CiliumPodIPPoolSelectorName defines the name for a selector matching CiliumPodIPPool resources.
+	CiliumPodIPPoolSelectorName = "CiliumPodIPPool"
```
Do we need this constant? We do have "CiliumPodIPPool" defined [here](https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/v2alpha1/register.go#L80 )

danehans replied:
@harsimran-pabla thanks for the pointer. I was following the existing constant pattern but I'll update to use your ref.

danehans replied:
@harsimran-pabla commit 51738e1 keeps the `CiliumPodIPPoolSelectorName` const but now refers to [the associated Kind](https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/v2alpha1/register.go#L80).

rastislavs replied:
I think it makes sense to keep this constant in the same file as the kubebuilder enum definition to make the API usage easier - referring to the associated Kind looks like a good middle ground 👍. Hopefully the kind won't ever change - as changing it would cause inconsistency with the kubebuilder enum values.

harsimran-pabla commented on pkg/bgpv1/manager/route_policy_reconciler.go:
```diff
@@ -186,8 +192,27 @@ func (r *RoutePolicyReconciler) pathAttributesToPolicy(attrs v2alpha1api.CiliumB
+			if cidrs, ok := localPools[pool.Name]; ok {
+				for _, cidr := range cidrs {
+					if cidr.Addr().Is4() {
+						prefix := int(pool.Spec.IPv4.MaskSize)
```
nit: should be prefixLen, same for v6

danehans replied:
@harsimran-pabla thanks for the review. Commit 677c3ca updates the var name to `prefixLen`.

christarazi reviewed (commented):
The K8s changes LGTM to me. Could you add justification in the commit msg? Without it, we lose potentially important context in git history if later down the line we bisect this commit.
